mutation_reader.cc

```diff
@@ -2294,3 +2294,432 @@ flat_mutation_reader make_compacting_reader(flat_mutation_reader source, gc_cloc
+                }
+
+                if (mf->is_end_of_partition()) {
+                    _galloping_reader->reader.pop_mutation_fragment();
+                    _all_readers.erase(_galloping_reader);
+                } else {
+                    if (_reader_queue->empty(mf->position())
+                            && (_peeked_readers.empty()
+                                    || _cmp(mf->position(), _peeked_readers.front()->reader.peek_buffer().position()) < 0)) {
+                        _current_batch.push_back(_galloping_reader->reader.pop_mutation_fragment());
+
+                        return make_ready_future<mutation_fragment_batch>(_current_batch);
+                    }
+
+                    // One of the existing readers won with the galloping reader,
+                    // or there is a yet unselected reader which possibly has a smaller position.
+                    // In either case we exit the galloping mode.
+
+                    _peeked_readers.emplace_back(_galloping_reader);
+                    boost::range::push_heap(_peeked_readers, _peeked_cmp);
+                }
+            } else {
+                // See comment in `peek_reader`.
+                _halted_readers.push_back(_galloping_reader);
+                boost::range::push_heap(_halted_readers, _halted_reader_cmp);
+            }
+
+            // The galloping reader has either been removed, halted, or lost with the other readers.
+            // Proceed with the normal path.
+            _galloping_reader = {};
+            _gallop_mode_hits = 0;
+            return (*this)(timeout);
+        });
+    }
+
+public:
+    clustering_order_reader_merger(
+            schema_ptr schema, reader_permit permit,
+            streamed_mutation::forwarding fwd_sm,
+            std::unique_ptr<position_reader_queue> reader_queue)
+        : _schema(std::move(schema)), _permit(std::move(permit))
+        , _cmp(*_schema)
+        , _reader_queue(std::move(reader_queue))
+        , _peeked_cmp(*_schema)
+        , _halted_reader_cmp(*_schema)
+        , _pr_end(fwd_sm == streamed_mutation::forwarding::yes
+                        ? position_in_partition_view::before_all_clustered_rows()
+                        : position_in_partition_view::after_all_clustered_rows())
+        , _should_return_partition_end(fwd_sm == streamed_mutation::forwarding::no)
+    {
+        // The first call to `_reader_queue::pop` uses `after_all_clustered_rows`
+        // so we obtain at least one reader; we will return this reader's `partition_start`
+        // as the first fragment.
+        auto rs = _reader_queue->pop(position_in_partition_view::after_all_clustered_rows());
+        if (rs.empty()) {
+            on_internal_error(mrlog,
+                    "clustering_order_reader_merger: no readers available."
+                    " This merger must be given at least one reader to work correctly.");
+        }
```

Comment:
I don't think this warrants an `on_internal_error()`. Just set EOS and done.

Reply:
Or `operator()()` can't handle no readers? This condition will also happen when reaching EOS genuinely. Callers are allowed (although it doesn't make sense) to call `fill_buffer()` on an EOS'd reader too.

Reply:
I wanted to guarantee the following:
_all merged readers return `partition_start` => the merger returns `partition_start`_.
But this can only be guaranteed if we have at least one reader. The merger could also construct a `partition_start` on its own, but it would need to know the partition key (initially that's what it did, but I wanted to simplify the reader and make it more efficient so now it obtains the `partition_start` from the readers).

Reply:
Won't removing this check just work and the reader will be in instant EOS?

Reply:
Yes, it will.

I actually added this check because my test catched that this reader when merging 0 readers will return a different sequence of fragments than one created by `flat_mutation_reader_from_mutations` for empty mutation: it will instantly return EOS, while the reader_from_mutations will return `partition_start` followed by `partition_end`.

Hence I modified the test so it doesn't check the 0 readers/empty mutation case, and added this check here.

Isn't returning EOS before `partition_start` wrong? Wouldn't it confuse cache populator like in https://github.com/scylladb/scylla/issues/3552?

Reply:
The reason is that an empty mutation != no data. En empty mutation contains exactly what you got: a partition start and end fragments.

Reply:
> Isn't returning EOS before partition_start wrong? Wouldn't it confuse cache populator like in #3552?

This is the problem of the higher layer, the combined reader also doesn't concern itself of how one of the many users react to no data. They were the one passing in 0 readers in the first place, they should deal with the result.